File: BackEnd/help.js
"use client"
// pages/index.js
import { useState, useEffect } from "react"
import Head from "next/head"
import {
  Bar,
  BarChart,
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts"
import { CSVLink } from "react-csv"

// API URL - change this to match your Flask server
const API_URL = "http://localhost:5000/api"

export default function Home() {
  // State for filters
  const [state, setState] = useState("")
  const [city, setCity] = useState("")
  const [businessType, setBusinessType] = useState("")
  const [companySize, setCompanySize] = useState("")
  const [filteredData, setFilteredData] = useState([])
  const [chartData, setChartData] = useState([])
  const [trendData, setTrendData] = useState([])
  const [dropdownOpen, setDropdownOpen] = useState({
    state: false,
    city: false,
    businessType: false,
    companySize: false,
  })
  const [alerts, setAlerts] = useState([])
  const [currentFee, setCurrentFee] = useState("")

  // State for filter options from API
  const [states, setStates] = useState([])
  const [cities, setCities] = useState([])
  const [businessTypes, setBusinessTypes] = useState([])
  const [companySizes, setCompanySizes] = useState([])
  
  // State for insights
  const [insights, setInsights] = useState({
    averageMedian: 0,
    feeRange: { min: 0, max: 0 },
    trendAnalysis: "Insufficient data to determine trend.",
    feeAnalysis: "No data available for comparison.",
    alerts: []
  })

  // Pagination state
  const [currentPage, setCurrentPage] = useState(1)
  const rowsPerPage = 5

  // Fetch filter options from API
  useEffect(() => {
    const fetchFilters = async () => {
      try {
        const response = await fetch(`${API_URL}/filters`)
        if (response.ok) {
          const data = await response.json()
          setStates(data.states)
          setCities(data.cities)
          setBusinessTypes(data.businessTypes)
          setCompanySizes(data.companySizes)
        } else {
          console.error("Failed to fetch filters:", response.statusText)
        }
      } catch (error) {
        console.error("Error fetching filters:", error)
      }
    }

    fetchFilters()
  }, [])

  // Fetch cities based on selected state
  useEffect(() => {
    const fetchCities = async () => {
      if (state) {
        try {
          const response = await fetch(`${API_URL}/filters?state=${state}`)
          if (response.ok) {
            const data = await response.json()
            setCities(data.cities)
          } else {
            console.error("Failed to fetch cities:", response.statusText)
          }
        } catch (error) {
          console.error("Error fetching cities:", error)
        }
      }
    }

    fetchCities()
    setCity("") // Reset city when state changes
  }, [state])

  // Toggle dropdown
  const toggle = (dropdown) => {
    // Close all other dropdowns when opening a new one
    const newState = {
      state: false,
      city: false,
      businessType: false,
      companySize: false,
    }

    newState[dropdown] = !dropdownOpen[dropdown]
    setDropdownOpen(newState)
  }

  // Fetch data based on filters
  useEffect(() => {
    const fetchData = async () => {
      try {
        // Build query params
        const params = new URLSearchParams()
        if (state) params.append("state", state)
        if (city) params.append("city", city)
        if (businessType) params.append("businessType", businessType)
        if (companySize) params.append("companySize", companySize)
        
        const response = await fetch(`${API_URL}/data?${params.toString()}`)
        if (response.ok) {
          const data = await response.json()
          setFilteredData(data)
          setCurrentPage(1) // Reset to first page when filters change
        } else {
          console.error("Failed to fetch data:", response.statusText)
        }
      } catch (error) {
        console.error("Error fetching data:", error)
      }
    }

    fetchData()
  }, [state, city, businessType, companySize])

  // Fetch chart data and insights
  useEffect(() => {
    const fetchChartData = async () => {
      try {
        // Build query params
        const params = new URLSearchParams()
        if (state) params.append("state", state)
        if (city) params.append("city", city)
        if (businessType) params.append("businessType", businessType)
        if (companySize) params.append("companySize", companySize)
        
        const response = await fetch(`${API_URL}/chart-data?${params.toString()}`)
        if (response.ok) {
          const data = await response.json()
          setChartData(data.barChart)
          setTrendData(data.lineChart)
        } else {
          console.error("Failed to fetch chart data:", response.statusText)
        }
      } catch (error) {
        console.error("Error fetching chart data:", error)
      }
    }

    const fetchInsights = async () => {
      try {
        // Build query params
        const params = new URLSearchParams()
        if (state) params.append("state", state)
        if (city) params.append("city", city)
        if (businessType) params.append("businessType", businessType)
        if (companySize) params.append("companySize", companySize)
        if (currentFee) params.append("currentFee", currentFee)
        
        const response = await fetch(`${API_URL}/insights?${params.toString()}`)
        if (response.ok) {
          const data = await response.json()
          setInsights(data)
          setAlerts(data.alerts)
        } else {
          console.error("Failed to fetch insights:", response.statusText)
        }
      } catch (error) {
        console.error("Error fetching insights:", error)
      }
    }

    fetchChartData()
    fetchInsights()
  }, [state, city, businessType, companySize, currentFee])

  // Reset all filters
  const resetFilters = () => {
    setState("")
    setCity("")
    setBusinessType("")
    setCompanySize("")
    setCurrentFee("")
    setCurrentPage(1)
  }

  // Handle option selection
  const handleSelect = (type, value) => {
    if (type === "state") {
      setState(value)
    } else if (type === "city") {
      setCity(value)
    } else if (type === "businessType") {
      setBusinessType(value)
    } else if (type === "companySize") {
      setCompanySize(value)
    }

    // Close the dropdown
    setDropdownOpen({
      ...dropdownOpen,
      [type]: false,
    })
  }
  
  // Pagination handlers
  const indexOfLastRow = currentPage * rowsPerPage
  const indexOfFirstRow = indexOfLastRow - rowsPerPage
  const currentRows = filteredData.slice(indexOfFirstRow, indexOfLastRow)
  const totalPages = Math.ceil(filteredData.length / rowsPerPage)
  
  const nextPage = () => {
    if (currentPage < totalPages) {
      setCurrentPage(currentPage + 1)
    }
  }
  
  const prevPage = () => {
    if (currentPage > 1) {
      setCurrentPage(currentPage - 1)
    }
  }

  // Custom dropdown styles
  const dropdownStyles = {
    menu: {
      maxHeight: "200px",
      overflowY: "auto",
      width: "100%",
      padding: "0",
      margin: "0",
      borderRadius: "4px",
      boxShadow: "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)",
      zIndex: 1000,
      position: "absolute",
      backgroundColor: "white",
    },
    item: {
      padding: "8px 12px",
      cursor: "pointer",
      borderBottom: "1px solid #f0f0f0",
      transition: "background-color 0.2s",
    },
    itemHover: {
      backgroundColor: "#f9fafb",
    },
  }

  return (
    <div className="container mx-auto px-4 py-8">
      <Head>
        <title>Fee Benchmark Dashboard</title>
        <meta name="description" content="Interactive fee benchmark dashboard" />
        <link rel="icon" href="/favicon.ico" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap/5.3.0/css/bootstrap.min.css" />
      </Head>

      <main>
        <h1 className="text-3xl font-bold mb-8 text-center">Fee Benchmark Dashboard</h1>

        {/* Filters Section */}
        <div className="bg-gray-100 p-6 rounded-lg mb-8">
          <div className="flex flex-wrap justify-between items-center mb-4">
            <h2 className="text-xl font-bold">Filters</h2>
            <button onClick={resetFilters} className="bg-blue-500 text-white px-4 py-2 rounded hover:bg-gray-600">
              Reset Filters
            </button>
          </div>

          <div className="grid grid-cols-1 md:grid-cols-4 gap-4">
            {/* State Filter */}
            <div>
              <label className="block mb-2 font-semibold">State</label>
              <div className="relative">
                <button
                  onClick={() => toggle("state")}
                  className="w-full text-left bg-white border rounded p-2 flex justify-between items-center"
                >
                  <span>{state || "Select State"}</span>
                  <span className="ml-2">▼</span>
                </button>
                {dropdownOpen.state && (
                  <div style={dropdownStyles.menu} className="mt-1 border">
                    <table className="w-full">
                      <tbody>
                        {states.map((st) => (
                          <tr
                            key={st}
                            onClick={() => handleSelect("state", st)}
                            className="hover:bg-gray-100 cursor-pointer"
                          >
                            <td className="p-2 border-b">{st}</td>
                          </tr>
                        ))}
                      </tbody>
                    </table>
                  </div>
                )}
              </div>
            </div>

            {/* City Filter */}
            <div>
              <label className="block mb-2 font-semibold">City</label>
              <div className="relative">
                <button
                  onClick={() => toggle("city")}
                  className="w-full text-left bg-white border rounded p-2 flex justify-between items-center"
                  disabled={!state}
                >
                  <span>{city || "Select City"}</span>
                  <span className="ml-2">▼</span>
                </button>
                {dropdownOpen.city && (
                  <div style={dropdownStyles.menu} className="mt-1 border">
                    <table className="w-full">
                      <tbody>
                        {cities.map((cty) => (
                          <tr
                            key={cty}
                            onClick={() => handleSelect("city", cty)}
                            className="hover:bg-gray-100 cursor-pointer"
                          >
                            <td className="p-2 border-b">{cty}</td>
                          </tr>
                        ))}
                      </tbody>
                    </table>
                  </div>
                )}
              </div>
            </div>

            {/* Business Type Filter */}
            <div>
              <label className="block mb-2 font-semibold">Business Type</label>
              <div className="relative">
                <button
                  onClick={() => toggle("businessType")}
                  className="w-full text-left bg-white border rounded p-2 flex justify-between items-center"
                >
                  <span>{businessType || "Select Business Type"}</span>
                  <span className="ml-2">▼</span>
                </button>
                {dropdownOpen.businessType && (
                  <div style={dropdownStyles.menu} className="mt-1 border">
                    <table className="w-full">
                      <tbody>
                        {businessTypes.map((type) => (
                          <tr
                            key={type}
                            onClick={() => handleSelect("businessType", type)}
                            className="hover:bg-gray-100 cursor-pointer"
                          >
                            <td className="p-2 border-b">{type}</td>
                          </tr>
                        ))}
                      </tbody>
                    </table>
                  </div>
                )}
              </div>
            </div>

            {/* Company Size Filter */}
            <div>
              <label className="block mb-2 font-semibold">Company Size</label>
              <div className="relative">
                <button
                  onClick={() => toggle("companySize")}
                  className="w-full text-left bg-white border rounded p-2 flex justify-between items-center"
                >
                  <span>{companySize || "Select Company Size"}</span>
                  <span className="ml-2">▼</span>
                </button>
                {dropdownOpen.companySize && (
                  <div style={dropdownStyles.menu} className="mt-1 border">
                    <table className="w-full">
                      <tbody>
                        {companySizes.map((size) => (
                          <tr
                            key={size}
                            onClick={() => handleSelect("companySize", size)}
                            className="hover:bg-gray-100 cursor-pointer"
                          >
                            <td className="p-2 border-b">{size} employees</td>
                          </tr>
                        ))}
                      </tbody>
                    </table>
                  </div>
                )}
              </div>
            </div>
          </div>

          {/* Current Fee Input */}
          <div className="mt-4">
            <label className="block mb-2 font-semibold">Your Current Fee ($)</label>
            <input
              type="number"
              value={currentFee}
              onChange={(e) => setCurrentFee(e.target.value)}
              className="w-full md:w-1/3 p-2 border rounded"
              placeholder="Enter your current fee"
            />
          </div>
        </div>

        {/* Alerts Section */}
        {alerts.length > 0 && (
          <div className="bg-red-100 border-l-4 border-red-500 text-red-700 p-4 mb-8" role="alert">
            <p className="font-bold">Alert</p>
            {alerts.map((alert, index) => (
              <p key={index}>{alert}</p>
            ))}
          </div>
        )}
        {/* Benchmark Visualization */}
        <div className="grid grid-cols-1 lg:grid-cols-2 gap-8 mb-8">
          {/* Bar Chart */}
          <div className="bg-white p-6 rounded-lg shadow">
            <h2 className="text-xl font-bold mb-4">Fee Benchmarks</h2>
            <div className="h-64">
              <ResponsiveContainer width="100%" height="100%">
                <BarChart data={chartData} margin={{ top: 20, right: 30, left: 20, bottom: 5 }}>
                  <CartesianGrid strokeDasharray="3 3" />
                  <XAxis dataKey="name" />
                  <YAxis label={{ value: "Fee ($)", angle: -90, position: "insideLeft" }} />
                  <Tooltip formatter={(value) => ["$" + value, "Fee"]} />
                  <Bar dataKey="value" fill="#4f46e5" />
                </BarChart>
              </ResponsiveContainer>
            </div>
          </div>

          {/* Line Chart for Trends */}
          <div className="bg-white p-6 rounded-lg shadow">
            <h2 className="text-xl font-bold mb-4">Fee Trends Over Time</h2>
            <div className="h-64">
              <ResponsiveContainer width="100%" height="100%">
                <LineChart data={trendData} margin={{ top: 20, right: 30, left: 20, bottom: 5 }}>
                  <CartesianGrid strokeDasharray="3 3" />
                  <XAxis dataKey="quarter" />
                  <YAxis label={{ value: "Fee ($)", angle: -90, position: "insideLeft" }} />
                  <Tooltip formatter={(value) => ["$" + value, "Fee"]} />
                  <Legend />
                  <Line type="monotone" dataKey="25th Percentile" stroke="#8884d8" activeDot={{ r: 8 }} />
                  <Line type="monotone" dataKey="Median" stroke="#82ca9d" />
                  <Line type="monotone" dataKey="75th Percentile" stroke="#ffc658" />
                </LineChart>
              </ResponsiveContainer>
            </div>
          </div>
        </div>

        {/* Data Table */}
        <div className="bg-white p-6 rounded-lg shadow mb-8">
          <div className="flex justify-between items-center mb-4">
            <h2 className="text-xl font-bold">Fee Benchmark Data</h2>
            <CSVLink
              data={filteredData}
              filename="fee_benchmarks.csv"
              className="bg-green-600 text-white px-4 py-2 rounded hover:bg-green-700"
            >
              Export to CSV
            </CSVLink>
          </div>
          <div className="overflow-x-auto">
            {/* Table container with fixed height for scrolling */}
            <div className="h-80 overflow-y-auto border rounded">
              <table className="min-w-full bg-white">
                <thead className="sticky top-0 bg-gray-100">
                  <tr>
                    <th className="py-2 px-4 border-b text-left">State</th>
                    <th className="py-2 px-4 border-b text-left">City</th>
                    <th className="py-2 px-4 border-b text-left">Business Type</th>
                    <th className="py-2 px-4 border-b text-left">Company Size</th>
                    <th className="py-2 px-4 border-b text-left">Quarter</th>
                    <th className="py-2 px-4 border-b text-right">25th Percentile</th>
                    <th className="py-2 px-4 border-b text-right">Median</th>
                    <th className="py-2 px-4 border-b text-right">75th Percentile</th>
                  </tr>
                </thead>
                <tbody>
                  {currentRows.map((item) => (
                    <tr key={item.id} className="hover:bg-gray-50">
                      <td className="py-2 px-4 border-b">{item.state}</td>
                      <td className="py-2 px-4 border-b">{item.city}</td>
                      <td className="py-2 px-4 border-b">{item.businessType}</td>
                      <td className="py-2 px-4 border-b">{item.companySize} employees</td>
                      <td className="py-2 px-4 border-b">{item.quarter}</td>
                      <td className="py-2 px-4 border-b text-right">${item.percentile25}</td>
                      <td className="py-2 px-4 border-b text-right">${item.median}</td>
                      <td className="py-2 px-4 border-b text-right">${item.percentile75}</td>
                    </tr>
                  ))}
                  {filteredData.length === 0 && (
                    <tr>
                      <td colSpan="8" className="py-4 text-center text-gray-500">
                        No data available with the selected filters. Please adjust your filters.
                      </td>
                    </tr>
                  )}
                </tbody>
              </table>
            </div>
            
            {/* Pagination Controls */}
            {filteredData.length > 0 && (
              <div className="flex justify-between items-center mt-4">
                <div className="text-sm text-gray-600">
                  Showing {indexOfFirstRow + 1} to {Math.min(indexOfLastRow, filteredData.length)} of {filteredData.length} entries
                </div>
                <div className="flex space-x-2">
                  <button 
                    onClick={prevPage} 
                    disabled={currentPage === 1}
                    className={`px-4 py-2 border rounded ${currentPage === 1 ? 'bg-gray-100 text-gray-400 cursor-not-allowed' : 'bg-white text-gray-700 hover:bg-gray-50'}`}
                  >
                    Previous
                  </button>
                  <span className="px-4 py-2 bg-gray-100 border rounded">
                    {currentPage} of {totalPages}
                  </span>
                  <button 
                    onClick={nextPage} 
                    disabled={currentPage === totalPages || totalPages === 0}
                    className={`px-4 py-2 border rounded ${currentPage === totalPages || totalPages === 0 ? 'bg-gray-100 text-gray-400 cursor-not-allowed' : 'bg-white text-gray-700 hover:bg-gray-50'}`}
                  >
                    Next
                  </button>
                </div>
              </div>
            )}
          </div>
        </div>

        {currentFee && Number.parseFloat(currentFee) > 0 && (
          <div className="bg-white p-6 rounded-lg shadow">
            <h2 className="text-xl font-bold mb-4">Data Insights</h2>
            {filteredData.length > 0 ? (
              <div>
                <p className="mb-2">
                  <strong>Average Median Fee:</strong> ${insights.averageMedian}
                </p>
                <p className="mb-2">
                  <strong>Fee Range:</strong> ${insights.feeRange.min} - ${insights.feeRange.max}
                </p>
                <p className="mb-2">
                  <strong>Trend Analysis:</strong> {insights.trendAnalysis}
                </p>

                {currentFee && (
                  <div className="mt-4 p-4 bg-gray-100 rounded">
                    <h3 className="font-bold mb-2">Your Fee Analysis</h3>
                    <p className="mb-2">{insights.feeAnalysis}</p>
                  </div>
                )}
              </div>
            ) : (
              <p className="text-gray-500">Select filters to view insights.</p>
            )}
          </div>
        )}
      </main>

      <footer className="mt-12 text-center text-gray-500">
        <p>© 2025 Fee Benchmark Dashboard</p>
      </footer>
    </div>
  )
}

File: BackEnd/help.py
from flask import Flask, jsonify, request
from flask_cors import CORS
import json

app = Flask(__name__)
CORS(app)  # Enable CORS for all routes

# Updated sample data with state and city fields instead of geography
sample_data = [
    {
        "id": 1,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2500,
        "median": 3200,
        "percentile75": 4100,
        "quarter": "Q1 2024",
    },
    {
        "id": 2,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2600,
        "median": 3300,
        "percentile75": 4200,
        "quarter": "Q2 2024",
    },
    {
        "id": 3,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2700,
        "median": 3400,
        "percentile75": 4300,
        "quarter": "Q3 2024",
    },
    {
        "id": 4,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Healthcare",
        "companySize": "150-200",
        "percentile25": 3100,
        "median": 3900,
        "percentile75": 4800,
        "quarter": "Q1 2024",
    },
    {
        "id": 5,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Healthcare",
        "companySize": "150-200",
        "percentile25": 3200,
        "median": 4000,
        "percentile75": 4900,
        "quarter": "Q2 2024",
    },
    {
        "id": 6,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Healthcare",
        "companySize": "150-200",
        "percentile25": 3300,
        "median": 4100,
        "percentile75": 5000,
        "quarter": "Q3 2024",
    },
    {
        "id": 7,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "200-250",
        "percentile25": 3200,
        "median": 4100,
        "percentile75": 5200,
        "quarter": "Q1 2024",
    },
    {
        "id": 8,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "200-250",
        "percentile25": 3300,
        "median": 4200,
        "percentile75": 5300,
        "quarter": "Q2 2024",
    },
    {
        "id": 9,
        "state": "Texas",
        "city": "Dallas",
        "businessType": "Educational Services",
        "companySize": "200-250",
        "percentile25": 3400,
        "median": 4300,
        "percentile75": 5400,
        "quarter": "Q3 2024",
    },
    {
        "id": 10,
        "state": "Texas",
        "city": "Houston",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2300,
        "median": 3000,
        "percentile75": 3800,
        "quarter": "Q1 2024",
    },
    {
        "id": 11,
        "state": "Texas",
        "city": "Houston",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2400,
        "median": 3100,
        "percentile75": 3900,
        "quarter": "Q2 2024",
    },
    {
        "id": 12,
        "state": "Texas",
        "city": "Houston",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 2500,
        "median": 3200,
        "percentile75": 4000,
        "quarter": "Q3 2024",
    },
    {
        "id": 13,
        "state": "California",
        "city": "Los Angeles",
        "businessType": "Educational Services",
        "companySize": "150-200",
        "percentile25": 3200,
        "median": 4100,
        "percentile75": 5200,
        "quarter": "Q1 2024",
    },
    {
        "id": 14,
        "state": "California",
        "city": "San Francisco",
        "businessType": "Healthcare",
        "companySize": "200-250",
        "percentile25": 4100,
        "median": 5200,
        "percentile75": 6300,
        "quarter": "Q2 2024",
    }
]


@app.route("/api/data", methods=["GET"])
def get_data():
    """
    Get all benchmark data, with optional filtering
    """
    # Extract query parameters
    state = request.args.get("state")
    city = request.args.get("city")
    business_type = request.args.get("businessType")
    company_size = request.args.get("companySize")
    
    # Start with all data
    filtered_data = sample_data.copy()
    
    # Apply filters if provided
    if state:
        filtered_data = [item for item in filtered_data if item["state"] == state]
    
    if city:
        filtered_data = [item for item in filtered_data if item["city"] == city]
    
    if business_type:
        filtered_data = [item for item in filtered_data if item["businessType"] == business_type]
    
    if company_size:
        filtered_data = [item for item in filtered_data if item["companySize"] == company_size]
    
    return jsonify(filtered_data)


@app.route("/api/filters", methods=["GET"])
def get_filters():
    """
    Get unique values for all filter options
    """
    # Extract unique values
    states = sorted(list(set(item["state"] for item in sample_data)))
    cities = sorted(list(set(item["city"] for item in sample_data)))
    business_types = sorted(list(set(item["businessType"] for item in sample_data)))
    company_sizes = sorted(list(set(item["companySize"] for item in sample_data)))
    
    filters = {
        "states": states,
        "cities": cities,
        "businessTypes": business_types,
        "companySizes": company_sizes
    }
    
    # Optional: Get cities specific to a state if state parameter is provided
    state_param = request.args.get("state")
    if state_param:
        state_cities = sorted(list(set(item["city"] for item in sample_data if item["state"] == state_param)))
        filters["cities"] = state_cities
    
    return jsonify(filters)


@app.route("/api/chart-data", methods=["GET"])
def get_chart_data():
    """
    Get data formatted for the charts
    """
    # Extract query parameters
    state = request.args.get("state")
    city = request.args.get("city")
    business_type = request.args.get("businessType")
    company_size = request.args.get("companySize")
    
    # Start with all data
    filtered_data = sample_data.copy()
    
    # Apply filters if provided
    if state:
        filtered_data = [item for item in filtered_data if item["state"] == state]
    
    if city:
        filtered_data = [item for item in filtered_data if item["city"] == city]
    
    if business_type:
        filtered_data = [item for item in filtered_data if item["businessType"] == business_type]
    
    if company_size:
        filtered_data = [item for item in filtered_data if item["companySize"] == company_size]
    
    # If no data after filtering, return empty response
    if not filtered_data:
        return jsonify({
            "barChart": [],
            "lineChart": []
        })
    
    # Get the latest quarter data for bar chart
    latest = sorted(filtered_data, key=lambda x: x["quarter"], reverse=True)[0]
    bar_chart = [
        {"name": "25th Percentile", "value": latest["percentile25"]},
        {"name": "Median", "value": latest["median"]},
        {"name": "75th Percentile", "value": latest["percentile75"]}
    ]
    
    # Prepare trend data for line chart
    quarters = sorted(list(set(item["quarter"] for item in filtered_data)))
    trends = []
    
    for quarter in quarters:
        quarter_data = next((item for item in filtered_data if item["quarter"] == quarter), None)
        if quarter_data:
            trends.append({
                "quarter": quarter,
                "25th Percentile": quarter_data["percentile25"],
                "Median": quarter_data["median"],
                "75th Percentile": quarter_data["percentile75"]
            })
    
    return jsonify({
        "barChart": bar_chart,
        "lineChart": trends
    })


@app.route("/api/insights", methods=["GET"])
def get_insights():
    """
    Get analytical insights based on filtered data
    """
    # Extract query parameters
    state = request.args.get("state")
    city = request.args.get("city")
    business_type = request.args.get("businessType")
    company_size = request.args.get("companySize")
    current_fee = request.args.get("currentFee")
    
    # Convert current_fee to float if provided
    if current_fee:
        try:
            current_fee = float(current_fee)
        except ValueError:
            current_fee = 0
    else:
        current_fee = 0
    
    # Start with all data
    filtered_data = sample_data.copy()
    
    # Apply filters if provided
    if state:
        filtered_data = [item for item in filtered_data if item["state"] == state]
    
    if city:
        filtered_data = [item for item in filtered_data if item["city"] == city]
    
    if business_type:
        filtered_data = [item for item in filtered_data if item["businessType"] == business_type]
    
    if company_size:
        filtered_data = [item for item in filtered_data if item["companySize"] == company_size]
    
    # If no data after filtering, return empty insights
    if not filtered_data:
        return jsonify({
            "averageMedian": 0,
            "feeRange": {"min": 0, "max": 0},
            "trendAnalysis": "Insufficient data to determine trend.",
            "feeAnalysis": "No data available for comparison.",
            "alerts": []
        })
    
    # Calculate insights
    avg_median = sum(item["median"] for item in filtered_data) / len(filtered_data)
    min_fee = min(item["percentile25"] for item in filtered_data)
    max_fee = max(item["percentile75"] for item in filtered_data)
    
    # Sort by quarter for trend analysis
    sorted_data = sorted(filtered_data, key=lambda x: x["quarter"])
    
    # Analyze trend if we have at least two quarters
    if len(sorted_data) > 1:
        first_median = sorted_data[0]["median"]
        last_median = sorted_data[-1]["median"]
        
        if last_median > first_median:
            trend_pct = ((last_median - first_median) / first_median) * 100
            trend_text = f"Fees have increased by {round(trend_pct)}% over the displayed period."
        else:
            trend_pct = ((first_median - last_median) / first_median) * 100
            trend_text = f"Fees have decreased by {round(trend_pct)}% over the displayed period."
    else:
        trend_text = "Insufficient data to determine trend."
    
    # Fee analysis based on current fee
    alerts = []
    fee_analysis = "No current fee provided for analysis."
    
    if current_fee > 0 and filtered_data:
        latest = sorted(filtered_data, key=lambda x: x["quarter"], reverse=True)[0]
        
        if current_fee < latest["percentile25"]:
            fee_analysis = "Your current fee is below the 25th percentile for your selected criteria."
            fee_status = "competitive"
        elif current_fee < latest["median"]:
            fee_analysis = "Your current fee is between the 25th percentile and median for your selected criteria."
            fee_status = "average"
        elif current_fee < latest["percentile75"]:
            fee_analysis = "Your current fee is between the median and 75th percentile for your selected criteria."
            fee_status = "above average"
        else:
            fee_analysis = "Your current fee is above the 75th percentile for your selected criteria."
            fee_status = "high"
            alerts.append(f"Warning: Your current fee of ${current_fee} is above the 75th percentile (${latest['percentile75']})")
    
    insights = {
        "averageMedian": round(avg_median),
        "feeRange": {"min": min_fee, "max": max_fee},
        "trendAnalysis": trend_text,
        "feeAnalysis": fee_analysis,
        "alerts": alerts
    }
    
    return jsonify(insights)


if __name__ == "__main__":
    app.run(debug=True, port=5000)